### PR TITLE
Add Pelagos provisioning service support

### PR DESCRIPTION
**Pull request by ryg-**
Add provisioning support via Pelagos provisioner https://github.com/ryg-/pelagos
Integration tests here pelagos/test_pelagos_teuthology/test_pelagos.py because depends on executed Pelagios service
provision/pelagos.py: added support of Pelagos provisioning project, interface is compatible with FOG provisioner
provision/__init.py: added processing of pelagos section in teuthology configuration and provisioner instantiation
lock/*: added Pelagos provisioner instantiation

Signed-off-by: Roman Grigorev <rgrigorev@suse.de>

### CODE REVIEW

**Comment by kshtsk** [comment]
ok to test

File: teuthology/provision/pelagos.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        self.log.error("do_request code %s text %s", resp.status_code, resp.text)
+        if not resp.ok and resp.text:
+            self.log.error("%s: %s", resp.status_code, resp.text)
+        if verify:
+            resp.raise_for_status()
+        return resp
+
+    def destroy(self):
+        """A no-op; we just leave idle nodes as-is"""
+        pass
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: extra destroy declaration?

File: teuthology/provision/pelagos.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        return
+
+    def is_task_active(self, task_url):
+        try:
+            status_response = self.do_request('', url=task_url, verify=False)
+        except HTTPError as err:
+            self.log.error("Task fail reason: " + err.reason)
+            if err.status_code == 404:
+                self.log.error(err.reason)
+                self.task_status_response = 'faield'
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: *failed

File: teuthology/provision/pelagos.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+def enabled(warn=False):
+    """
+    Check for required Pelagos settings
+
+    :param warn: Whether or not to log a message containing unset parameters
+    :returns: True if they are present; False if they are not
+    """
+    conf = config.get(config_section, dict())
+    params = ['endpoint', 'machine_types']
+    unset = [param for param in params if not conf.get(param)]
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: Could you change "param" to "_" for readability, too many 'param-param-pa-pams' in one line :-)

File: teuthology/provision/pelagos.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    :returns: The list of Pelagos-configured machine types. An empty list if Pelagos is
+              not configured.
+    """
+    if not enabled():
+        return []
+    conf = config.get(config_section, dict())
+    types = conf.get('machine_types', '')
+    if not isinstance(types, list):
+        types = types.split(',')
+    return [type_ for type_ in types if type_]
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: type_ => _

File: teuthology/provision/pelagos.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+    :returns: The list of Pelagos-configured machine types. An empty list if Pelagos is
+              not configured.
+    """
+    if not enabled():
+        return []
+    conf = config.get(config_section, dict())
+    types = conf.get('machine_types', '')
+    if not isinstance(types, list):
+        types = types.split(',')
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: I would do here:
```
    if not isinstance(types, list):
        types = [_ for _ in types.split(',') if _]
```

File: teuthology/provision/pelagos.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    return [type_ for type_ in types if type_]
+
+
+class Pelagos(object):
+
+    def __init__(self, name, os_type, os_version):
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: just wondering, are you going to support arch?

File: teuthology/provision/pelagos.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            location = response.headers.get('Location')
+            self.log.info("Waiting for deploy to finish")
+            self.log.info("Observe location: " + location)
+            time.sleep(2)
+            with safe_while(sleep=15, tries=60) as proceed:
+                while proceed():
+                    if not self.is_task_active(location):
+                        break
+        except Exception as e:
+            # TODO implement cancel task
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: You already have dummy function, why not to call it from here?

File: teuthology/provision/pelagos.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        :param data: Optional JSON data to submit with the request
+        :param method: The HTTP method to use for the request (default: 'GET')
+        :param verify: Whether or not to raise an exception if the request is
+                       unsuccessful (default: True)
+        :returns: A requests.models.Response object
+        """
+        prepared_url = config.pelagos['endpoint'] + url_suffix
+        if url != '':
+            prepared_url = url
+        self.log.info("Connect to :" + prepared_url)
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: (style) remove spaces before ':'

File: teuthology/provision/pelagos.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        if data is not None:
+            self.log.info("Send data  :" + str(data))
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: (style) the same as above, no need spaces before ':'

File: teuthology/provision/pelagos.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            status_response = self.do_request('', url=task_url, verify=False)
+        except HTTPError as err:
+            self.log.error("Task fail reason: " + err.reason)
+            if err.status_code == 404:
+                self.log.error(err.reason)
+                self.task_status_response = 'faield'
+                return False
+            else:
+                raise HTTPError(err.code, err.reason)
+        self.log.info("Response code:[" +
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: (style), could you please remove square brackets, and use quotes, also use '%s' with loging function, like:
```
log.info("Response code: '%s'", status_response.status_code)
```
This is applicable to all standard info, warning, error methods.

File: teuthology/provision/pelagos.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+                          str(status_response.status_code) + "]")
+        self.task_status_response = status_response
+        if status_response.status_code == 202:
+            self.log.info("Status response:[" +
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: Same here?

File: teuthology/provision/pelagos.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        split_uri = re.search(r'(\w*)@(.+)', name)
+        if split_uri is not None:
+            self.name = split_uri.groups()[1]
+        else:
+            self.name = name
+
+        self.os_type = os_type
+        self.os_version = os_version
+        self.os_name = os_type + "-" + os_version
+        self.log = log.getChild(self.name)
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: Why do you need this? Why not using it directly `log.method()`?

File: teuthology/provision/pelagos.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+from teuthology.config import config
+from teuthology.contextutil import safe_while
+from teuthology.exceptions import MaxWhileTries
+from teuthology import misc
+
+log = logging.getLogger(__name__)
+config_section = 'pelagos'
+
+
+def enabled(warn=False):
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: Why don't you move it to the class as a member method?

File: teuthology/provision/pelagos.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    unset = [param for param in params if not conf.get(param)]
+    if unset and warn:
+        log.warn(
+            "Pelagos is disabled; set the following config options to enable: %s",
+            ' '.join(unset),
+        )
+    return (unset == [])
+
+
+def get_types():
```
Comment (kshtsk) [inline teuthology/provision/pelagos.py]: Why don't you move it to the class as a member method?

File: teuthology/provision/pelagos.py (hunk 6 above)
Comment (kshtsk) [reply teuthology/provision/pelagos.py]: looks like fog does not have it either

File: teuthology/provision/pelagos.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+import teuthology.orchestra
+from teuthology.config import config
+from teuthology.contextutil import safe_while
+from teuthology.exceptions import MaxWhileTries
+from teuthology import misc
+
+log = logging.getLogger(__name__)
+config_section = 'pelagos'
+
```
Comment (toabctl) [inline teuthology/provision/pelagos.py]: Could we have at least some very basic docs (here in this file or in the teuthology rst docs) about the available config options in this section?

File: teuthology/provision/__init__.py [hunk 16]
```diff
@@ -2,14 +2,18 @@
 
 import teuthology.lock.query
 from teuthology.misc import decanonicalize_hostname, get_distro, get_distro_version
+from teuthology.config import config
+
```
Comment (kshtsk) [inline teuthology/provision/__init__.py]: remove extra empty line

File: teuthology/provision/__init__.py [hunk 17]
```diff
@@ -2,14 +2,18 @@
...
 
 from teuthology.provision import cloud
 from teuthology.provision import downburst
 from teuthology.provision import fog
 from teuthology.provision import openstack
+from teuthology.provision import pelagos
 import os
 
 
+
```
Comment (kshtsk) [inline teuthology/provision/__init__.py]: remove extra empty line

File: teuthology/provision/__init__.py (hunk 16 above)
Comment (ryg-) [reply teuthology/provision/__init__.py]: done

File: teuthology/provision/__init__.py (hunk 17 above)
Comment (ryg-) [reply teuthology/provision/__init__.py]: done

**Comment by ryg-** [comment]
@susebot run deploy

**Comment by toabctl** [comment]
@ryg- why was this closed?

**Comment by kshtsk** [comment]
> @ryg- why was this closed?

@toabctl it was closed by an accident.

**Pull request closed without merging**